### ISSUE TITLE
I tried to run a librispeech recipe but a word error rate remains very large. 

### ISSUE THREAD

**Issue by ken57** [freewym/espresso#27]
#### What is your question?
I tried to run a librispeech recipe (examples/asr_librispeech/run.sh) but a word error rate remains very large(around 100%) in "Stage 8: Model Training" in spite of 30 epochs. 
I think a cause is that an execution environment is different.

#### What's your environment?
My environment is as follows.

 - fairseq Version (e.g., 1.0 or master): 0.9.0
 - PyTorch Version (e.g., 1.0): 1.4.0
 - OS (e.g., Linux): Ubuntu 18.04.3 LTS
 - How you installed fairseq (`pip`, source): pip
 - Python version: 3.6.5
 - CUDA/cuDNN version: 10.0.130 / libcudnn.so.7.3.0
 - GPU models and configuration: Tesla V100-SXM2-16GB

```
$ python collect_env.py 
Collecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.0

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: Tesla V100-SXM2-16GB
Nvidia driver version: 440.33.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.3.0

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] torch==1.4.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2020.0                      166  
[conda] mkl-service               2.3.0            py36he904b0f_0  
[conda] mkl_fft                   1.0.15           py36ha843d7b_0  
[conda] mkl_random                1.1.0            py36hd6b4f25_0  
[conda] pytorch                   1.4.0           py3.6_cuda10.0.130_cudnn7.6.3_0    pytorch
[conda] torch                     1.4.0                    pypi_0    pypi
```

A commit hash of the espresso is f933e8cfff38a20bd7ea76dd4cb3a1fa809eab83.
An output log is as follows, but I can't find any problem.

```
2020-03-08 22:27:56 | INFO | espresso.criterions.label_smoothed_cross_entropy_v2 | sample REF: I SEE I MUST GET SETTLED QUICKLY SO THAT I SHALL HAVE THE POWER TO RESTRAIN YOU THEY ROLLICKED FORTH THEN AND BOUGHT SEVERAL THINGS A BIG STEAMER RUG FOR THE CAR A PAIR OF LONG GRAY MOCHA GLOVES TO MATCH THE HAND BAG A SILK UMBRELLA
2020-03-08 22:27:56 | INFO | espresso.criterions.label_smoothed_cross_entropy_v2 | sample PRD: AND HAVE THAT' BE A IN AND I I CAN BE TO PLEASURE OF GET THE I ARE UPED AND AND THE THERE THE OF ANDPIECEGERER ANDG AND THE SHIPS FEW OF SHOES WHITE HAIRSACKS WHICH MAN OF PAIR HANDKERCHIEF
2020-03-08 23:13:04 | INFO | valid | epoch 025 | valid on 'valid' subset | loss 6.804 | nll_loss 5.856 | wer 109.669 | cer 100.424 | ppl 57.94 | wps 822.3 | wpb 715.6 | bsz 29.1 | num_updates 414000 | best_wer 96.7413
2020-03-08 23:13:27 | INFO | fairseq.checkpoint_utils | saved checkpoint exp/lstm/checkpoint_25_414000.pt (epoch 25 @ 414000 updates, score 109.6687) (writing took 23.222585418028757 seconds)
2020-03-08 23:13:28 | INFO | espresso.criterions.label_smoothed_cross_entropy_v2 | sample REF: CONTINUED DUNCAN SPEAKING SLOWLY AND USING THE SIMPLEST FRENCH OF WHICH HE WAS THE MASTER TO BELIEVE THAT NONE OF THIS WISE AND BRAVE NATION UNDERSTAND THE LANGUAGE THAT THE GRAND MONARQUE USES WHEN HE TALKS TO HIS CHILDREN
2020-03-08 23:13:28 | INFO | espresso.criterions.label_smoothed_cross_entropy_v2 | sample PRD: AND THECAN WITH IN AND IING THE WORDSST WAY LANGUAGE THE HE WAS THE MOST OF BE THAT HE OF THE WAS AND IN MAN COULDS LANGUAGE OF HE WORLDESTITTSS HE ISS TO THE PEOPLE
2020-03-08 23:23:27 | INFO | train | epoch 025:  15999 / 16601 loss=6.624, nll_loss=5.649, ppl=50.18, wps=537.4, ups=0.75, wpb=716.6, bsz=16.9, num_updates=414424, lr=1e-05, gnorm=0.399, clip=0, oom=0, train_wall=13155, wall=554591
2020-03-08 23:34:58 | INFO | train | epoch 025 | loss 6.624 | nll_loss 5.649 | ppl 50.18 | wps 539.9 | ups 0.75 | wpb 716.2 | bsz 16.9 | num_updates 415025 | lr 1e-05 | gnorm 0.4 | clip 0 | oom 0 | train_wall 13641 | wall 555281
2020-03-08 23:37:49 | INFO | valid | epoch 025 | valid on 'valid' subset | loss 6.803 | nll_loss 5.856 | wer 108.198 | cer 99.5523 | ppl 57.92 | wps 822.3 | wpb 715.6 | bsz 29.1 | num_updates 415025 | best_wer 96.7413
2020-03-08 23:38:12 | INFO | fairseq.checkpoint_utils | saved checkpoint exp/lstm/checkpoint25.pt (epoch 25 @ 415025 updates, score 108.1984) (writing took 23.870150407077745 seconds)
```

Because the librispeech is very large dataset, I struggle with debugging.
Could you give me any hint?

I think if the espresso has a recipe for a small dataset like the an4 recipe in the espnet, a trial run is easier.
Do you have any plan to implement a recipe for a small dataset?

thanks.

**Comment by freewym** [comment]
Does increasing `--curriculum` help? like 2?
My experiment on Libripspeech was done with 2GPUs, and actually I didn't try a setup with 1 GPU. Normally the WER became reasonable after training for 3-4 epochs. And when I encountered divergence, increasing `--curriculum` usually helps.

I think WSJ is small enough. you can see  a reasonable decoding result in an hour.

**Comment by ken57** [comment]
Thank you for your quick response.

> Does increasing --curriculum help? like 2?

Now, I trying to run the librispeech recepi with `--curriculum 2`.

```
  CUDA_VISIBLE_DEVICES=$free_gpu speech_train.py data --task speech_recognition_espresso --seed 1 --user-dir espresso \
    --log-interval 4000 --log-format simple --print-training-sample-interval 2000 \
    --num-workers 0 --max-tokens 26000 --max-sentences 24 --curriculum 2 \
    --valid-subset $valid_subset --max-sentences-valid 48 --ddp-backend no_c10d \
```

> Normally the WER became reasonable after training for 3-4 epochs.

This is very useful information!
I waited a week for training 30 epochs every time...

> I think WSJ is small enough. you can see a reasonable decoding result within an hour.

Unfortunately, I don't have WSJ.
I will consider purchasing WSJ corpus.

**Comment by freewym** [comment]
Also, independently, try `--warmup-updates` at 1000 or so.

**Comment by ken57** [comment]
`--curriculum 2` is ineffective.
```
2020-03-11 10:47:07 | INFO | valid | epoch 001 | valid on 'valid' subset | loss 8.016 | nll_loss 7.285 | wer 149.992 | cer 130.43 | ppl 155.99 | wps 864.8 | wpb 715.6 | bsz 29.1 | num_updates 3000
2020-03-11 11:50:48 | INFO | valid | epoch 001 | valid on 'valid' subset | loss 7.704 | nll_loss 6.925 | wer 170.511 | cer 145.011 | ppl 121.5 | wps 820.3 | wpb 715.6 | bsz 29.1 | num_updates 6000 | best_wer 149.992
2020-03-11 13:01:24 | INFO | valid | epoch 001 | valid on 'valid' subset | loss 7.56 | nll_loss 6.758 | wer 185.179 | cer 155.713 | ppl 108.25 | wps 782.6 | wpb 715.6 | bsz 29.1 | num_updates 9000 | best_wer 149.992
2020-03-11 14:15:25 | INFO | valid | epoch 001 | valid on 'valid' subset | loss 7.48 | nll_loss 6.667 | wer 306.926 | cer 249.097 | ppl 101.63 | wps 573.2 | wpb 715.6 | bsz 29.1 | num_updates 12000 | best_wer 149.992
2020-03-11 15:34:08 | INFO | valid | epoch 001 | valid on 'valid' subset | loss 7.431 | nll_loss 6.605 | wer 310.311 | cer 251.265 | ppl 97.37 | wps 570.2 | wpb 715.6 | bsz 29.1 | num_updates 15000 | best_wer 149.992
2020-03-11 16:10:29 | INFO | valid | epoch 001 | valid on 'valid' subset | loss 7.5 | nll_loss 6.693 | wer 328.229 | cer 265.346 | ppl 103.45 | wps 559 | wpb 715.6 | bsz 29.1 | num_updates 16601 | best_wer 149.992
2020-03-11 16:32:51 | INFO | valid | epoch 002 | valid on 'valid' subset | loss 7.467 | nll_loss 6.631 | wer 122.194 | cer 111.54 | ppl 99.12 | wps 1033.3 | wpb 715.6 | bsz 29.1 | num_updates 18000 | best_wer 122.194
2020-03-11 17:39:50 | INFO | valid | epoch 002 | valid on 'valid' subset | loss 7.378 | nll_loss 6.543 | wer 194.071 | cer 162.303 | ppl 93.27 | wps 775.5 | wpb 715.6 | bsz 29.1 | num_updates 21000 | best_wer 122.194
2020-03-11 18:45:06 | INFO | valid | epoch 002 | valid on 'valid' subset | loss 7.349 | nll_loss 6.504 | wer 225.539 | cer 185.997 | ppl 90.75 | wps 723.6 | wpb 715.6 | bsz 29.1 | num_updates 24000 | best_wer 122.194
2020-03-11 19:57:21 | INFO | valid | epoch 002 | valid on 'valid' subset | loss 7.328 | nll_loss 6.481 | wer 351.798 | cer 284.003 | ppl 89.32 | wps 541.8 | wpb 715.6 | bsz 29.1 | num_updates 27000 | best_wer 122.194
2020-03-11 21:12:03 | INFO | valid | epoch 002 | valid on 'valid' subset | loss 7.334 | nll_loss 6.487 | wer 294.114 | cer 238.816 | ppl 89.69 | wps 567.6 | wpb 715.6 | bsz 29.1 | num_updates 30000 | best_wer 122.194
2020-03-11 22:22:21 | INFO | valid | epoch 002 | valid on 'valid' subset | loss 7.369 | nll_loss 6.53 | wer 294.079 | cer 238.502 | ppl 92.41 | wps 584.6 | wpb 715.6 | bsz 29.1 | num_updates 33000 | best_wer 122.194
2020-03-11 22:31:35 | INFO | valid | epoch 002 | valid on 'valid' subset | loss 7.388 | nll_loss 6.554 | wer 489.041 | cer 393.247 | ppl 93.94 | wps 382.4 | wpb 715.6 | bsz 29.1 | num_updates 33202 | best_wer 122.194
2020-03-11 23:30:30 | INFO | valid | epoch 003 | valid on 'valid' subset | loss 7.12 | nll_loss 6.243 | wer 100.662 | cer 98.4372 | ppl 75.75 | wps 1235.5 | wpb 715.6 | bsz 29.1 | num_updates 36000 | best_wer 100.662
2020-03-12 00:37:03 | INFO | valid | epoch 003 | valid on 'valid' subset | loss 7.075 | nll_loss 6.187 | wer 109.216 | cer 100.849 | ppl 72.88 | wps 854.2 | wpb 715.6 | bsz 29.1 | num_updates 39000 | best_wer 100.662
2020-03-12 01:46:37 | INFO | valid | epoch 003 | valid on 'valid' subset | loss 7.037 | nll_loss 6.15 | wer 120.478 | cer 105.133 | ppl 71.02 | wps 820.3 | wpb 715.6 | bsz 29.1 | num_updates 42000 | best_wer 100.662
2020-03-12 02:54:53 | INFO | valid | epoch 003 | valid on 'valid' subset | loss 7.022 | nll_loss 6.137 | wer 115.448 | cer 101.733 | ppl 70.38 | wps 816.7 | wpb 715.6 | bsz 29.1 | num_updates 45000 | best_wer 100.662
2020-03-12 04:03:52 | INFO | valid | epoch 003 | valid on 'valid' subset | loss 7.015 | nll_loss 6.132 | wer 115.536 | cer 102.435 | ppl 70.12 | wps 840.6 | wpb 715.6 | bsz 29.1 | num_updates 48000 | best_wer 100.662
2020-03-12 04:45:13 | INFO | valid | epoch 003 | valid on 'valid' subset | loss 7.006 | nll_loss 6.118 | wer 112.624 | cer 101.08 | ppl 69.44 | wps 840.8 | wpb 715.6 | bsz 29.1 | num_updates 49803 | best_wer 100.662
2020-03-12 05:11:48 | INFO | valid | epoch 004 | valid on 'valid' subset | loss 7.014 | nll_loss 6.111 | wer 104.643 | cer 98.2152 | ppl 69.1 | wps 923.8 | wpb 715.6 | bsz 29.1 | num_updates 51000 | best_wer 100.662
2020-03-12 06:17:08 | INFO | valid | epoch 004 | valid on 'valid' subset | loss 6.994 | nll_loss 6.099 | wer 121.358 | cer 106.504 | ppl 68.56 | wps 841.8 | wpb 715.6 | bsz 29.1 | num_updates 54000 | best_wer 100.662
2020-03-12 07:22:38 | INFO | valid | epoch 004 | valid on 'valid' subset | loss 6.981 | nll_loss 6.089 | wer 109.311 | cer 101.827 | ppl 68.08 | wps 860.1 | wpb 715.6 | bsz 29.1 | num_updates 57000 | best_wer 100.662
2020-03-12 08:29:24 | INFO | valid | epoch 004 | valid on 'valid' subset | loss 6.983 | nll_loss 6.082 | wer 105.883 | cer 97.888 | ppl 67.76 | wps 887.9 | wpb 715.6 | bsz 29.1 | num_updates 60000 | best_wer 100.662
2020-03-12 09:34:37 | INFO | valid | epoch 004 | valid on 'valid' subset | loss 6.968 | nll_loss 6.074 | wer 110.421 | cer 100.249 | ppl 67.36 | wps 858.8 | wpb 715.6 | bsz 29.1 | num_updates 63000 | best_wer 100.662
2020-03-12 10:41:42 | INFO | valid | epoch 004 | valid on 'valid' subset | loss 6.957 | nll_loss 6.061 | wer 108.38 | cer 98.9368 | ppl 66.75 | wps 867.1 | wpb 715.6 | bsz 29.1 | num_updates 66000 | best_wer 100.662
```

Now, I trying to run the librispeech recepi with `--warmup-updates 1000`

**Comment by ken57** [comment]
--warmup-updates 1000 seems to be effective...
I will watch the progress for a while.
```
2020-03-12 21:29:31 | INFO | valid | epoch 002 | valid on 'valid' subset | loss 2.506 | nll_loss 0.967 | wer 20.0408 | cer 23.8809 | ppl 1.96 | wps 783 | wpb 715.6 | bsz 29.1 | num_updates 27000 | best_wer 20.0408
2020-03-12 22:51:12 | INFO | valid | epoch 002 | valid on 'valid' subset | loss 2.471 | nll_loss 0.938 | wer 18.5515 | cer 21.9329 | ppl 1.92 | wps 811.4 | wpb 715.6 | bsz 29.1 | num_updates 30000 | best_wer 18.5515
2020-03-13 00:06:58 | INFO | valid | epoch 002 | valid on 'valid' subset | loss 2.639 | nll_loss 1.125 | wer 22.9786 | cer 26.584 | ppl 2.18 | wps 792.7 | wpb 715.6 | bsz 29.1 | num_updates 33000 | best_wer 18.5515
2020-03-13 00:13:07 | INFO | valid | epoch 002 | valid on 'valid' subset | loss 2.588 | nll_loss 1.062 | wer 21.1865 | cer 24.6818 | ppl 2.09 | wps 792.1 | wpb 715.6 | bsz 29.1 | num_updates 33202 | best_wer 18.5515
```

**Comment by ken57** [comment]
Training of "Stage 8: Model Training" is finished.
The WER is shown below on 30 epochs in Stage 8.
```
2020-03-21 01:21:03 | INFO | valid | epoch 030 | valid on 'valid' subset | loss 2.052 | nll_loss 0.53 | wer 7.9468 | cer 9.8567 | ppl 1.44 | wps 805.3 | wpb 715.6 | bsz 29.1 | num_updates 492000 | best_wer 7.9013
2020-03-21 02:29:12 | INFO | valid | epoch 030 | valid on 'valid' subset | loss 2.052 | nll_loss 0.531 | wer 7.9041 | cer 9.7903 | ppl 1.44 | wps 807.9 | wpb 715.6 | bsz 29.1 | num_updates 495000 | best_wer 7.9013
2020-03-21 03:36:36 | INFO | valid | epoch 030 | valid on 'valid' subset | loss 2.052 | nll_loss 0.53 | wer 7.926 | cer 9.8224 | ppl 1.44 | wps 809.4 | wpb 715.6 | bsz 29.1 | num_updates 498000 | best_wer 7.9013
2020-03-21 03:40:13 | INFO | valid | epoch 030 | valid on 'valid' subset | loss 2.052 | nll_loss 0.53 | wer 7.9193 | cer 9.8155 | ppl 1.44 | wps 806.9 | wpb 715.6 | bsz 29.1 | num_updates 498030 | best_wer 7.9013
```
Is this expected result?

**Comment by freewym** [comment]
I think so. My `best_wer` number with 2 GPUs is 7.95

**Comment by ken57** [comment]
My question was resolved!
Thank you very much for kind support.
